---
title: "DesertAnalysis_SstarSimulatedData "
output: html_notebook
---

```{r Load Packages}
library(ggplot2)
library(dplyr)
library(data.table)
library(tidyr)
library(scales)
library(grid)
library(gridExtra)
library(stringr)

require(bit64)
```

```{r LOAD DATA}
#######################
# ADMIX DATA

outputdir <- '~/DATALab/SimulatedDemographic/Sstar/test/multi_sample/0.05_pct/'
 mdl <- 'Tenn_nonAfr'
 admix <- 'n1_0.05_n2_0.0'
 maxchrm <- 10
 dt <- data.table(NULL)
 for(i in seq(1,as.numeric(maxchrm),by = 1)){
    print(paste0('Loading chromosome number: ',i))
    infile <- paste0(outputdir, '/RegionFiles/', mdl, "_",i,'_',admix,".windowcalc_out.gz")
    dat <- read.table(gzfile(infile), header=TRUE, na.strings=c("NA", "None",'.'), as.is = TRUE)
    dat <- as.data.table(dat)
    dt <- rbind(dt, dat)
 }
admix_5pct.dt <- dt


########################
# ADMIX PVALUES




admix_5pct.match_pvals <- dt
#######################
# NULL DATA

outputdir <- '~/DATALab/SimulatedDemographic/Sstar/test/multi_sample/null/'
mdl <- 'Tenn_nonAfr'
admix <- 'n1_0.0_n2_0.0'
sim_chrms <- fread(paste0(outputdir,"Tenn.chr_list"))
dt <- data.table(NULL)
for( i in seq(1,1000,by = 1)){
#for( i in seq(1,nrow(sim_chrms),by = 1)){
    c <- sim_chrms[i][[1]]
    print(paste0('Loading chromosome number: ',c, i))
    infile <- paste0(outputdir, '/RegionFiles/', mdl, "_",c,'_',admix,".windowcalc_out.gz")
    dat <- read.table(gzfile(infile), header=TRUE, na.strings=c("NA", "None",'.'), as.is = TRUE)
    dat <- as.data.table(dat)
    dt <- rbind(dt, dat)
}
null.5k.dt <- dt
###########################
```


```{r DEFINE DATASETS}
admix.dt <- admix_5pct.dt
null.dt <- null.10k.dt
```


```{r MODIFY DATA TABLES}
admix.dt[, filter := (s_star <= 0)]
admix.dt[,sstarpval_region_ind_snps := 0.00]  # make new column of double type
admix.dt[,sstarpval_region_ind_snps := NA]  # replace values with NA
admix.dt[filter==FALSE, "sstarpval_region_ind_snps" := 1]  ## initiate with value 1 if window passed filters'


null.dt[, filter := (s_star <= 0)]
null.dt[,"sstarpval" := 0.00]  # make new column of double type
null.dt[,"sstarpval" := NA]  # replace values with NA
null.dt[filter==FALSE, "sstarpval" := 1]  ## initiate with value 1 if window passed filters
```


```{r ECDF GENERATION AND PVALUE FUNCTION}
#############
# GENERATE ECDFS
for( i in sort(unique(as.numeric(null.dt$n_region_ind_snps)))){
  if(i>0){
  print(i)
  nam <- paste0('null.f.region_ind.', i, '.ecdf')
  if(nrow(filter(null.dt, s_star!=0, n_region_ind_snps==i))>0){
    assign(nam, ecdf(filter(null.dt, s_star!=0, n_region_ind_snps==i)$s_star))
    max_snps_ecdf <- i
  }}
}
#############

#############
# FUNCTION: Calculate S*-pvalue from ecdf
estimate.pval.ecdf.region_ind.fn <- function(X, max_snps){
  s_star <- as.numeric(X[["s_star"]])
  n_snps <- as.numeric(X[["n_region_ind_snps"]])
  if (n_snps==0){
    X[["sstarpval_region_ind_snps"]] <- NA
    } else if (n_snps<=max_snps) {
    if( exists(paste0("null.f.region_ind.",n_snps,".ecdf")) ){
      ecdf.fn <- match.fun(paste0("null.f.region_ind.",n_snps,".ecdf"))
      s_star_pval <- 1-ecdf.fn(s_star)
      X[["sstarpval_region_ind_snps"]] <- round(x = s_star_pval, digits = 4)
      }
    } else if (n_snps>max_snps) {
    ecdf.fn <- match.fun(paste0("null.f.region_ind.",max_snps,".ecdf"))
    s_star_pval <- 1-ecdf.fn(s_star)
    X[["sstarpval_region_ind_snps"]] <- round(x = s_star_pval, digits = 4)
    }
  return(X)
}
##############


####################
# Apply ecdf function to admixture data

max_snps_ecdf <- 

out <- as.data.table(t(apply(X = admix.dt,MARGIN = 1,FUN = estimate.pval.ecdf.region_ind.fn, max_snps=244)))

```


```{r ASSIGN S* HAPLOTYPES}
####################
# Assign S* haplotype
req.snp.frac <- 0.8
out$s_star_hap_1 <- (as.numeric(out$n_s_star_snps_hap1) / as.numeric(out$num_s_star_snps)) >= req.snp.frac
out$s_star_hap_2 <- (as.numeric(out$n_s_star_snps_hap2) / as.numeric(out$num_s_star_snps)) >= req.snp.frac 

out[,chrom:=as.numeric(chrom)]
out[,winstart:=as.numeric(winstart)]
out[,winend:=as.numeric(winend)]
out[,hap_1_s_start:=as.numeric(hap_1_s_start)]
out[,hap_1_s_end:=(hap_1_s_end)]
out[,hap_2_s_start:=as.numeric(hap_2_s_start)]
out[,hap_2_s_end:=(hap_2_s_end)]
out[,s_star:=as.numeric(s_star)]
out[,sstarpval_region_ind_snps:=as.numeric(sstarpval_region_ind_snps)]
out[,s_star_hap_1:=as.character(s_star_hap_1)]
out[,s_star_hap_2:=as.character(s_star_hap_2)]


hap_1 = out[s_star_hap_1==TRUE & s_star_hap_2==FALSE]
hap_2 = out[s_star_hap_1==FALSE & s_star_hap_2==TRUE]
hap_12 = out[s_star_hap_1==TRUE & s_star_hap_2==TRUE]
hap_NA = out[is.na(s_star_hap_1) | is.na(s_star_hap_2)]

hap_1[,haplotype:=1]
hap_2[,haplotype:=2]
hap_12_1  = data.table(hap_12[,haplotype:=1])
hap_12_2  = data.table(hap_12[,haplotype:=2])
hap_NA_1  = data.table(hap_NA[,haplotype:=1])
hap_NA_2  = data.table(hap_NA[,haplotype:=2])

out = rbind(
  hap_1,
  hap_2,
  hap_12_1,
  hap_12_2,
  hap_NA_1,
  hap_NA_2
)

out[,msp_ID:=paste0(ind_id,'_',chrom,'_',haplotype)]
out = out  %>% arrange(chrom, winstart, winend)
```


```{r ASSIGN MATCH PVALUES}
admix.match_pvals <- admix_5pct.N_MH_Der.isc_1.match_pvals

admix.pvals <- out %>% left_join((admix.match_pvals), by = c('msp_ID', 'winstart', 'winend','haplotype')) %>% as.data.table()

```


```{r VARIABLES FOR OUTPUT FILES}

outputdir <- '~/DATALab/SimulatedDemographic/Sstar/test/multi_sample/0.05_pct/'  
mdl <- 'Tenn_nonAfr'
admix <- 'n1_0.05_n2_0.0'
spval <- 0.01
matchpval <- 0.01
dt <- admix.pvals

```


```{r WRITE OUTPUT TABLES, FILTERED FOR S* AND MATCH PVALUES}
####################
# FILTERED FOR S*PVALUE AND MATCHPVALUE
dat_1 <- dt %>%
    filter(filter=='FALSE') %>%
    filter(sstarpval_region_ind_snps<=spval) %>%
    filter(match_pvalue<=matchpval) %>%
    filter(haplotype==1) %>%
    select(msp_ID, winstart, winend) %>%
    #select(msp_ID, hap_1_s_start, hap_1_s_end) %>%
    setnames(c('msp_ID','start','end')) %>%
    as.data.table()
  
dat_2 <- dt %>%
    filter(filter=='FALSE') %>%
    filter(sstarpval_region_ind_snps<=spval) %>%
    filter(match_pvalue<=matchpval) %>%
    filter(haplotype==2) %>%
    select(msp_ID, winstart, winend) %>%
    #select(msp_ID, hap_2_s_start, hap_2_s_end) %>%
    setnames(c('msp_ID','start','end')) %>%
    as.data.table()
  
dat.bed <- rbind(dat_1,dat_2)
  
options(scipen=10)
dat.bed %<>% mutate(start=str_trim(as.character(start),side = "both")) %>% mutate(end=str_trim(as.character(end), side = "both"))
write.table(x = dat.bed,
          file = paste0(outputdir,'test.sstar_sig_',spval,'.match_sig_N_',matchpval,'.isc_1','.bed'),
          quote = FALSE,
          sep = '\t',
          row.names = FALSE,
          col.names = FALSE)
options(scipen=0)
```


```{r WRITE OUTPUT TABLES, ALL S* AND MATCH PVALUES}
####################
# ALL WINDOWS
dat_1 <- dt %>%
    filter(filter=='FALSE') %>%
    filter(haplotype==1) %>%
    select(msp_ID, winstart, winend, sstarpval_region_ind_snps, match_pvalue) %>%
    #select(msp_ID, hap_1_s_start, hap_1_s_end, sstarpval_region_ind_snps, match_pvalue) %>%
    setnames(c('msp_ID','start','end', 'sstarpval_region_ind_snps', 'match_pvalue')) %>%
    as.data.table()
  
dat_2 <- dt %>%
    filter(filter=='FALSE') %>%
    filter(haplotype==2) %>%
    select(msp_ID, winstart, winend, sstarpval_region_ind_snps,match_pvalue) %>%
    #select(msp_ID, hap_2_s_start, hap_2_s_end, sstarpval_region_ind_snps,match_pvalue) %>%
    setnames(c('msp_ID','start','end', 'sstarpval_region_ind_snps', 'match_pvalue')) %>%
    as.data.table()
  
dat.bed <- rbind(dat_1,dat_2)
  
options(scipen=10)
dat.bed %<>% mutate(start=str_trim(as.character(start),side = "both")) %>% mutate(end=str_trim(as.character(end), side = "both"))
write.table(x = dat.bed,
          file = paste0(outputdir,'test.sstar_sig_','ALL','.match_sig_N_MH_','ALL','.isc_1','.bed'),
          quote = FALSE,
          sep = '\t',
          row.names = FALSE,
          col.names = FALSE)
options(scipen=0)
```

